awx_collection/plugins/modules/tower_credential_type.py

```diff
@@ -127,53 +115,45 @@ def main():
         state=dict(choices=['present', 'absent'], default='present'),
     )
 
-    module = TowerModule(
-        argument_spec=argument_spec,
-        supports_check_mode=False
-    )
+    # Create a module for ourselves
+    module = TowerModule(argument_spec=argument_spec, supports_check_mode=True)
 
+    # Extract our parameters
     name = module.params.get('name')
+    new_name = None
     kind = module.params.get('kind')
     state = module.params.get('state')
 
-    json_output = {'credential_type': name, 'state': state}
-
-    tower_auth = tower_auth_config(module)
-    with settings.runtime_values(**tower_auth):
-        tower_check_mode(module)
-        credential_type_res = tower_cli.get_resource('credential_type')
-
-        params = {}
-        params['name'] = name
-        params['kind'] = kind
-        params['managed_by_tower'] = False
-
-        if module.params.get('description'):
-            params['description'] = module.params.get('description')
-
-        if module.params.get('inputs'):
-            params['inputs'] = module.params.get('inputs')
-
-        if module.params.get('injectors'):
-            params['injectors'] = module.params.get('injectors')
-
-        try:
-            if state == 'present':
-                params['create_on_missing'] = True
-                result = credential_type_res.modify(**params)
-                json_output['id'] = result['id']
-            elif state == 'absent':
-                params['fail_on_missing'] = False
-                result = credential_type_res.delete(**params)
-
-        except (exc.ConnectionError, exc.BadRequest, exc.AuthError) as excinfo:
-            module.fail_json(
-                msg='Failed to update credential type: {0}'.format(excinfo),
-                changed=False
-            )
-
-    json_output['changed'] = result['changed']
-    module.exit_json(**json_output)
+    # These will be passed into the create/updates
+    credential_type_params = {
+        'name': new_name if new_name else name,
+        'kind': kind,
+        'managed_by_tower': False,
+    }
+    if module.params.get('description'):
+        credential_type_params['description'] = module.params.get('description')
+    if module.params.get('inputs'):
+        credential_type_params['inputs'] = module.params.get('inputs')
+    if module.params.get('injectors'):
+        credential_type_params['injectors'] = module.params.get('injectors')
+
+    # Attempt to look up credential_type based on the provided name
+    credential_type = module.get_one('credential_types', **{
+        'data': {
+            'name': name,
+        }
+    })
+
+    # Add entries to json_output to match old module
+    module.json_output['credential_type'] = name
+    module.json_output['state'] = state
```

Comment:
Wouldn't we prefer `state` to be part of the returned data? If you run a task in check mode, then you might make decisions based on what the state is, and it would make sense to report this even if it was given as a parameter. In that case, we should lose this line and have `delete_if_needed` and `create_or_update_if_needed` apply the state field.